packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -100,14 +118,42 @@ if (!ExecutionEnvironment.canUseDOM) {
   let previousFrameTime = 33;
   let activeFrameTime = 33;
 
-  const frameDeadlineObject = {
+  // When a callback is scheduled, we register it by adding it's id to this
+  // object.
+  // If the user calls 'cIC' with the id of that callback, it will be
+  // unregistered by removing the id from this object.
+  // Then we skip calling any callback which is not registered.
+  // This means cancelling is an O(1) time complexity instead of O(n).
+  const registeredCallbackIds: {[number]: boolean} = {};
+
+  const frameDeadlineObject: Deadline = {
     didTimeout: false,
     timeRemaining() {
       const remaining = frameDeadline - now();
       return remaining > 0 ? remaining : 0;
     },
   };
 
+  const safelyCallScheduledCallback = function(callback, callbackId) {
+    if (!registeredCallbackIds[callbackId]) {
+      // ignore cancelled callbacks
+      return;
+    }
+    isCurrentlyRunningCallback = true;
+    try {
+      callback(frameDeadlineObject);
+      delete registeredCallbackIds[callbackId];
+      isCurrentlyRunningCallback = false;
+    } catch (e) {
+      delete registeredCallbackIds[callbackId];
+      isCurrentlyRunningCallback = false;
+      // Still throw it, but not in this frame.
+      setTimeout(() => {
```

Comment:
Not sure I'm satisfied with this method of handling errors. We shouldn't swallow them, but don't want to throw when we may be in the middle of a queue of callbacks.
Exposing an 'onError' handler eventually maybe? Or taking an 'onError' option.